src/utils/typey-type.ts

```diff
@@ -0,0 +1,188 @@
+      true,
+    ];
+  }
+};
+
+// const migratePersonalDictionariesV1ToV2 = function (personalDictionaries, dirtyFlag) {
+//   if (personalDictionaries.v && personalDictionaries.v === '1') {
+//     let opts = {};
+//     let dictsWithMetadata = personalDictionaries.dicts.map(dict => [dict[0],dict[1],opts]);
+//     personalDictionaries = {v:'2',dicts:dictsWithMetadata};
+//     dirtyFlag = true;
+//   }
+
+//   return [personalDictionaries, dirtyFlag];
+// }
+
+const runAllPersonalDictionariesMigrations = function(
+  personalDictionaries: LocalStoragePersonalDictionariesUnknownVersion,
+  dirtyFlag: boolean
+): [LocalStoragePersonalDictionariesV1 | null, boolean, string | null] {
+  let error = null;
+  try {
+    [personalDictionaries, dirtyFlag] = migratePersonalDictionariesV0ToV1(
+      personalDictionaries,
+      dirtyFlag
+    );
+    // [personalDictionaries, dirtyFlag] = migratePersonalDictionariesV1ToV2(personalDictionaries, dirtyFlag);
+  } catch (exception) {
+    return [null, false, "Personal dictionaries found in local storage are either missing a version number or are not in valid version 0 format for migrating to version 1."];
+  }
+  return [personalDictionaries, dirtyFlag, error];
+};
+
+function migratePersonalDictionariesV(personalDictionaries: LocalStoragePersonalDictionariesUnknownVersion): [LocalStoragePersonalDictionariesV1 | null, string | null] {
+  let dirtyFlag = false;
+  let error: string | null = null;
+  let maybePersonalDictionaries: null | LocalStoragePersonalDictionariesUnknownVersion = personalDictionaries;
+
+  [maybePersonalDictionaries, dirtyFlag, error] =
+    runAllPersonalDictionariesMigrations(
+      maybePersonalDictionaries,
+      dirtyFlag,
+    );
+
+  if (maybePersonalDictionaries !== null && dirtyFlag && error === null) {
+    writePersonalPreferences("personalDictionaries", maybePersonalDictionaries);
+  }
+
+  return [maybePersonalDictionaries, error];
+}
+
+function loadPersonalDictionariesFromLocalStorage() {
+  try {
+    if (window.localStorage) {
+      let versionedDictionaries = null;
+      if (window.localStorage.getItem("personalDictionaries")) {
+        versionedDictionaries = JSON.parse(
+          window.localStorage.getItem("personalDictionaries") ?? "null"
+        );
+      }
+      if (versionedDictionaries === null) {
+        return null;
+      }
+
+      let errorMessage = null;
+      [versionedDictionaries, errorMessage] = migratePersonalDictionariesV(
+        versionedDictionaries
+      );
+      if (errorMessage) {
+        console.error(errorMessage);
+        return null;
+      }
+
+      let areDictsSomewhatValid =
+        versionedDictionaries !== null &&
+        versionedDictionaries.v &&
+        versionedDictionaries.v === "1" &&
+        versionedDictionaries.dicts &&
+        versionedDictionaries.dicts[0] &&
+        versionedDictionaries.dicts[0][1];
+      if (areDictsSomewhatValid) {
+        return versionedDictionaries?.["dicts"] ?? null;
```

Comment:
Adds `?.` and `?? null` to appease TS. In practice, `?.` should be no different to before because we already checked `versionedDictionaries.dicts` in the `if` conditional and if by some horror it's somehow changed since that check and returns `undefined` then we use `?? null` to explicitly return `null` so that `loadPersonalDictionariesFromLocalStorage()` always returns `PersonalDictionaryNameAndContents[] | null` and not maybe `undefined`.